rfcs/20190115-dense-attention.md

```diff
@@ -0,0 +1,506 @@
+# Attention for Dense networks on Keras
+
+| Status        | Proposed                                                                   |
+:-------------- |:-------------------------------------------------------------------------- |
+| **Author(s)** | Georgios Roumpos (roumposg@google.com)                                     |
+| **Sponsors**  | Karmel Allison (karmel@google.com), Francois Chollet (fchollet@google.com) |
+| **Updated**   | 2019-01-15                                                                 |
+
+## Objective and Motivation
+
+Recently people have had success using the Attention mechanism in dense layers,
+e.g. CNN+Attention or Transformer networks. Some examples are the
+["Attention is all you need"](https://arxiv.org/abs/1706.03762) paper, and
+models in
+[semantic text similarity](https://ai.googleblog.com/2018/05/advances-in-semantic-textual-similarity.html).
+`tf.keras.layers` is the recommended way to build models in tensorflow, but it
+does not have a layer for attention that works with CNN/DNN networks. We would
+like to contribute this capability.
+
+Keras is an API spec that can be implemented across different languages and
+backends, and `tf.keras` is a particular implementation of that spec. This
+document contains code examples for `tensorflow`, but the same API should work
+everywhere.
+
+### Recurrent Neural Networks
+
+Although not the primary focus of this proposal, the Attention layers in this
+proposal work with some configurations of RNN networks.
+Namely, when users create
+[tf.keras.layers.LSTM](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM)
+with `return_sequences=True`, the rest works the same way as CNN.
+
+Unfortunately, this technique does not cover sequence-to-sequence RNN models.
```

Comment:
By "this technique" do you mean your proposal?

If so I think your proposal should support masking, as it's very common.

Reply:
I see. This is a valid use case. Let me spend some more time to better understand this case and get back to you, thanks.

Reply:
I thought about this, and I don't think this can be implemented as a manual mask or a composable layer. It needs to be a feature of the layer.

I added a couple of options about how this can be implemented in the "Self-attention" section. Let me know what you think, thanks.